### ISSUE TITLE
ts_window start/end dates

### ISSUE THREAD

**Issue by DonPipo85** [christophsax/tsbox#63]
Salute,

wie kann ich ts_window(time_series, start = actual_q, end = q_end) die folgenden zwei Argumente übergeben?
actual_q <- c(2018,1)
q_end <- c(2019,4)
Ich möchte nicht manuell die Daten als String eingeben müssen, i.e. 
actual_q <- "2018-01-01"
q_end <- "2019-10-01"
LG Philipp

**Comment by christophsax** [comment]
I really would like to have the same input as `window`, but there are some issues with them. Summarizing my thoughts here: What you currently can do:

``` r
library(tsbox)
ts_window(AirPassengers, start = 1956)
#>      Jan Feb Mar Apr May Jun Jul Aug Sep Oct Nov Dec
#> 1956 284 277 317 313 318 374 413 405 355 306 271 306
#> 1957 315 301 356 348 355 422 465 467 404 347 305 336
#> 1958 340 318 362 348 363 435 491 505 404 359 310 337
#> 1959 360 342 406 396 420 472 548 559 463 407 362 405
#> 1960 417 391 419 461 472 535 622 606 508 461 390 432
```

which is just the same as:

    ts_window(AirPassengers, start = "1956-01-01")

Length 2 Frequencies
--------------------

The problem with length 2 inputs is that they are **frequency dependent**, and ts-boxable objects can be of mixed frequencies.

``` r
ap.quarter <- ts_frequency(AirPassengers, "quarter", sum)
ap.quarter
#>      Qtr1 Qtr2 Qtr3 Qtr4
#> 1949  362  385  432  341
#> 1950  382  409  498  387
#> 1951  473  513  582  474
#> 1952  544  582  681  557
#> 1953  628  707  773  592
#> 1954  627  725  854  661
#> 1955  742  854 1023  789
#> 1956  878 1005 1173  883
#> 1957  972 1125 1336  988
#> 1958 1020 1146 1400 1006
#> 1959 1108 1288 1570 1174
#> 1960 1227 1468 1736 1283
```

Because frequencies are mixed, this returns a data frame

``` r
mixed <- ts_c(ap.quarter, AirPassengers)
head(mixed, 15)
#>            id       time value
#> 1  ap.quarter 1949-01-01   362
#> 2  ap.quarter 1949-04-01   385
#> 3  ap.quarter 1949-07-01   432
#> 4  ap.quarter 1949-10-01   341
#> 5  ap.quarter 1950-01-01   382
#> 6  ap.quarter 1950-04-01   409
#> 7  ap.quarter 1950-07-01   498
#> 8  ap.quarter 1950-10-01   387
#> 9  ap.quarter 1951-01-01   473
#> 10 ap.quarter 1951-04-01   513
#> 11 ap.quarter 1951-07-01   582
#> 12 ap.quarter 1951-10-01   474
#> 13 ap.quarter 1952-01-01   544
#> 14 ap.quarter 1952-04-01   582
#> 15 ap.quarter 1952-07-01   681
```

Which can be forced back to `ts`:

``` r
head(ts_ts(mixed), 15)
#>       ap.quarter AirPassengers
#>  [1,]        362           112
#>  [2,]         NA           118
#>  [3,]         NA           132
#>  [4,]        385           129
#>  [5,]         NA           121
#>  [6,]         NA           135
#>  [7,]        432           148
#>  [8,]         NA           148
#>  [9,]         NA           136
#> [10,]        341           119
#> [11,]         NA           104
#> [12,]         NA           118
#> [13,]        382           115
#> [14,]         NA           126
#> [15,]         NA           141
```

The above still works:

``` r
ts_start(ts_window(mixed, start = "1956-01-01"))
#> [1] "1956-01-01"
```

This is also clear:

``` r
ts_start(ts_window(mixed, start = "1956-07-01"))
#> [1] "1956-07-01"
```

But what would be the meaning of this?

``` r
ts_window(mixed, start = c(1956, 3))
#> Error: numeric date input must be of length 1
```

**Variant 1:** My first reading was that this should give the 3rd quarter of `ap.quarter` and the 3rd month of `AirPassengers`, so the series would start at a different point in time.

The downside of this is that `window` and `ts_window` will work differently when applied to a mixed frequency `ts` object: Here, `ap.quarter` would start in Q3, `AirPassengers` in March:

``` r
ts_window(ts_ts(mixed), start = c(1956, 3))
#> Error: numeric date input must be of length 1
```

While here, both series will start in in March:

``` r
head(window(ts_ts(mixed), start = c(1956, 3)))
#>      ap.quarter AirPassengers
#> [1,]         NA           317
#> [2,]       1005           313
#> [3,]         NA           318
#> [4,]         NA           374
#> [5,]       1173           413
#> [6,]         NA           405
```

**Variant 2:** Or, since this is really a `"ts"` thing, we just do the same as what `window` does.

Downside is that adding a high frequency series to an object changes the behavior of `c(1956, 3)`.

I think this downside is more serious than the first one.

**Variant 3:** Not supporting them. In both 1 and 2, I would tell users not to use them anyway.

Decimal Fraction Frequencies
----------------------------

The other way to define start in `window` is to use decimal fractions. As they are not frequency dependent, they are easy to implement.

``` r
ts_start(window(AirPassengers, start = 1956.5))
#> [1] "1956-07-01"
ts_start(window(ap.quarter, start = 1956.5))
#> [1] "1956-07-01"
```

Long Story short: I agree it would be nice to support the `window` way of specifying start and end. The length 2 way is a bit awkward for mixed frequency objects, and I am somewhere between variant 1 and variant 3.

**Comment by christophsax** [comment]
If you really need it, why not use `window`?

**Comment by christophsax** [comment]
I guess we will not implement length 2 vectors, because they are terrible with mixed frequencies. Dates (and times) are at the core of tsbox, and it is just natural that we use them here as well. 

To reduce resemblance to base `window`, we probably go for `ts_span`, which is also shorter:

```r
ts_span(AirPassengers, from = "1956-3")
```
Thanks to anytime, dates can be entered without much typing.

We will also be able to do:

```r
ts_span(AirPassengers, from = "-1 year")
```
which gives us the latest available year.

or 

```r
ts_span(AirPassengers, from = "-100000 sec")
```
which gives us the latest 100000 seconds.

or:
```r
ts_span(AirPassengers, from = -5)
```
which returns the last 5 observation. Because this depends on the frequency, it will fail for mixed frequencies.

Objections?

Will be implemented soon.

**Comment by christophsax** [comment]
See `?ts_span`. As described above, but with argnames `start` and `end`.
